BaseUnit.cpp

```diff
@@ -0,0 +1,24 @@
+#include "BaseUnit.h"
+
+BaseUnit::BaseUnit(const std::string& nm, int hp, const int dmg) : Name{ nm }, HP{hp}, DMG{dmg}
+{
+}
+
+//decreasing the HP of the suffering unit, return with false if it dies
+bool BaseUnit::gotHit(int dam)
+{
+	HP = HP - dam;
+	if (HP <= 0)
+	{
+		HP = 0;
+		return false;
+	}
+		
+	return true;
+}
+
+//Show current HP and DMG
+const std::string BaseUnit::showStats() const
```

Comment:
nem sikerült a ctrl+c ctrl+v, úgyhogy még egyszer: `const std::string& getName() const { return Name; }`
összetett adattagot referenciaként adjuk vissza, hogy ne másolódjon feleslegesen, az érték szerint visszaadott stringnél nm sok értelme lenne a constnak

Reply:
Megpróbáltam így is. A Name-t tényleg vissza tudom adni referenciaként, ezt javítani is fogom.
De showStats esetén olyan változót kénék vissza adni, amit abban a függvényben hozok létre. Ha erre hivatkozok referenciával, akkor exceptiont kapok, mert kimegy a scope-ból.
Ezt most egy köztes std::string stats tagváltozóval oldottam meg, de így elveszíti a függvény a const tulajdonságát, mert módosítja egy adattag értékét

Reply:
Ezekre amúgy hogyan kellene válaszolnom?
Írtam ide és automatikusan reviewer lettem

Reply:
Így kell válaszolni, nem baj, ha emiatt reviewernek jelöl meg.
Bocsánat, lehet kicsit félreérthető helyre írtam a kommentet, vagy inkább 2 kommentben kellett volna.
Az egyik megjegyzés a `getName`re vonatkozott, hogy adjon vissza referenciát. A másik, hogy a `showStats`nál az, hogy constot ad vissza, annak nincs értelme. Viszont ott természetesen nem referenciát kell visszaadni, mert újonnan létrehozott objektummal tér vissza. Ezt a stats tagváltozót csináld vissza.

Reply:
Én is beszállok a beszélgetésbe egy kérdés erejéig. Tehát ha jól értem, ez a const-olás és referenciaként visszaadás csak olyan függvényeknél lehet és/vagy van értelme, ahol a függvény scope-jában olyan adattag szerepel return értéknek, ami a scope-on kívül is változatlanul szerepel (azaz egy adattagja az objektumnak)?
Pl. a mi esetünkben a getName-nél simán mehet a const és referencia szerinti visszaadás, mert olyan értéket ad vissza változatlanul ami konkrétan az objektum adattagja egy az egyben (feltételezem ez miatt próbáljuk meg elkerülni hogy az érték feleslegesen lemásolódjon)? Illetve a showStats-nál kénytelenek vagyunk elfelejteni a const-ot és a referencia szerinti értékvisszaadást, mivel egy módosított és csak scope-on belüli létező értékkel tér vissza (ami feltételezésem szerint lemásolandó, hogy ne legyen exception)?
Azaz, ez helyett a sor helyett.:
const std::string& BaseUnit::showStats()
Maradhat ez a sor.:
std::string BaseUnit::showStats()
És a getName pedig maradhat const + referencia szerint visszatérő?

Reply:
Így van, ahogy írtad, @kalmar-adam. De azért leírom a saját szavaimmal is, hogy egy kicsit pontosítsam.
Ha a getName referenciát ad vissza, akkor a hívás helyén is azt a memóriaterületet lehet, használni, ahol az objektum adattagja található. Ezért kell a const, hogy ne lehessen ezen keresztül megváltoztatni. Ha érték szerint adná vissza, akkor egy másolatot hozna létre a stacken, és ezt használná a hívó. Mivel ezzel nem tudja módosítani az eredetit, nem kell, hogy const legyen.
A showStats pedig egy új objektumot hoz létre, amit visszaad eredményül. Mivel lokális változóként hozza létre, ahogy írtad is, a függvényhívás végén megszűnik, ezért erre a memóriaterületre nem szabad hivatkozást visszaadni (se referenciát, se pointert). Nem az exception elkerüléséért, hanem mert érvénytelen memóriaterületre hivatkozna (use after return error). Muszáj érték szerint visszaadni, hogy átmásolódjon a hívó scope-jába. (Valójában az RVO - Return Value Optimization miatt ezt a másolást általában kikerüli a fordító.) A másik megoldás az lenne, hogy dinamikusan létrehozni az objektumot, akkor nem szűnik meg a függvény végén, de akkor meg a hívónak kéne gondoskodnia a felszabadításról, ami feleslegesen bonyolítaná, ezért ez nem jó megoldás. Azt akkor kell, ha utána hosszútávon szükség van az objektumra.

Reply:
@oliverosz Köszönöm így már értem. Most már látom, hogy az eredeti problémám az volt, hogy nem vettem észre a referencia jelet az első review-nál getName-nél, onnan eszkalálódott tovább az egész.
Megcsináltam a kért módosításokat. Használjam a re-request review gombot itt a githubon ahogy eddig, vagy tag-eljem meg Mátét, hogy ossza ki újra a feladatot, ahogy a levélben kérte?